src/test/java/io/cdap/plugin/batch/source/FTPBatchSourceTest.java

```diff
@@ -162,4 +182,121 @@ public void testSFTPPathWithSystemPropertiesAndSpecialCharactersInAuth() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    FakeFtpServer server = ftpSetup();
+    config.configuration("ftp://user:password@invalid_server:21", "");
+    config.validate(collector);
+    server.stop();
+    Assert.assertTrue(collector.getValidationFailures().size() > 0);
+  }
+
+  @Test
+  public void testInvalidUsernamePasswordFTPPathConnection() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    FakeFtpServer server = ftpSetup();
+    config.configuration("ftp://wronguser:wrongpassword@localhost:21", "");
+    config.validate(collector);
+    server.stop();
+    Assert.assertTrue(collector.getValidationFailures().size() > 0);
+  }
+
+  @Test
+  public void testValidFTPPathConnection() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    FakeFtpServer server = ftpSetup();
+    config.configuration("ftp://user:password@localhost:21", "");
+    config.validate(collector);
+    server.stop();
+    Assert.assertEquals(0, collector.getValidationFailures().size());
+  }
+
+  @Test
+  public void testInvalidServerSFTPPathConnection() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    try {
+      SshServer server = sftpSetup();
+      config.configuration("sftp://user:password@invalid_server:22", "");
+      config.validate(collector);
+      server.stop();
+      Assert.assertTrue(collector.getValidationFailures().size() > 0);
+    } catch (Exception e) {
+      // do nothing
+    }
+  }
+
+  @Test
+  public void testInvalidUsernamePasswordSFTPPathConnection() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    try {
+      SshServer server = sftpSetup();
+      config.configuration("sftp://wronguser:wrongpassword@localhost:22", "");
+      config.validate(collector);
+      server.stop();
+      Assert.assertTrue(collector.getValidationFailures().size() > 0);
+    } catch (Exception e) {
+      // do nothing
+    }
+  }
+
+  @Test
+  public void testValidSFTPPathConnection() {
+    FailureCollector collector = new MockFailureCollector();
+    FTPBatchSource.FTPBatchSourceConfig config = new FTPBatchSource.FTPBatchSourceConfig();
+    try {
+      SshServer server = sftpSetup();
+      config.configuration("sftp://user:password@localhost:22", "");
+      config.validate(collector);
+      server.stop();
+      Assert.assertEquals(0, collector.getValidationFailures().size());
+    } catch (Exception e) {
+      // do nothing
+    }
+  }
+
+  public static FakeFtpServer ftpSetup() {
```

Comment:
If you annotate methods with @BeforeClass/@AfterClass they will be called before/after test suite is executed.
In addition to setup methods, you can also add teardown methods to stop the servers